### ISSUE TITLE
Why tacotron use 256-dim character embeddings rather than one-hot vector for characters?

### ISSUE THREAD

**Issue by zyj008** [keithito/tacotron#206]
Hi, I am confused that why tacotron use 256-dim character embeddings rather than one-hot vector for characters. The total number of characters are small. What are the benefits of character embedding compared to one-hot?
Thank you.

**Comment by Pyrestone** [comment]
Hi, I'm not the author, but I can answer your question:

Mathematically, A one-hot-Vector followed by a matrix multiplication (i.e. Dense Layer) is equivalent to word/character embeddings, although, since one-hot vectors are sparse, it is simply more efficient to take a row out of the embedding matrix than to multiply said embedding matrix with the corresponding one-hot vector. This performance difference is larger for higher vocabulary sizes and might not be noticable for such small matrices, but it is still existant, and therefore good/respectable style to implement it this way.

Secondly, when keeping the full dataset as Input/Output pairs in RAM, the very sparse one-hot vectors take up much more space in RAM than the indexes of each character which are handed to the embedding layer. Again, this is probably not the issue since the size of the wav files and spectrograms should be far larger, but again, it's a little more efficient and therefore good style.

**TLDR:** In this case the Character embedding serves two purposes
* Slightly Increased Compute efficiency
* Much Increased Memory efficiency of the input data (although not that relevant since the output data is probably much larger in comparison)

**Comment by phypan11** [comment]
Why should it use size 256, while there are only < 100 characters? Isn't the size 32 or so enough?
